tests/integration/openapi3-express.spec.ts

```diff
@@ -0,0 +1,154 @@
+
+        expect(body.fields['body.floatValue'].message).to.equal('Invalid float error message.');
+        expect(body.fields['body.floatValue'].value).to.equal(bodyModel.floatValue);
+        expect(body.fields['body.doubleValue'].message).to.equal('Invalid double error message.');
+        expect(body.fields['body.doubleValue'].value).to.equal(bodyModel.doubleValue);
+        expect(body.fields['body.intValue'].message).to.equal('invalid integer number');
+        expect(body.fields['body.intValue'].value).to.equal(bodyModel.intValue);
+        expect(body.fields['body.longValue'].message).to.equal('Custom Required long number.');
+        expect(body.fields['body.longValue'].value).to.equal(bodyModel.longValue);
+        expect(body.fields['body.booleanValue'].message).to.equal('invalid boolean value');
+        expect(body.fields['body.booleanValue'].value).to.equal(bodyModel.booleanValue);
+
+        expect(body.fields['body.dateValue'].message).to.equal('invalid ISO 8601 date format, i.e. YYYY-MM-DD');
+        expect(body.fields['body.dateValue'].value).to.equal(bodyModel.dateValue);
+        expect(body.fields['body.datetimeValue'].message).to.equal('invalid ISO 8601 datetime format, i.e. YYYY-MM-DDTHH:mm:ss');
+        expect(body.fields['body.datetimeValue'].value).to.equal(bodyModel.datetimeValue);
+
+        expect(body.fields['body.numberMax10'].message).to.equal('max 10');
+        expect(body.fields['body.numberMax10'].value).to.equal(bodyModel.numberMax10);
+        expect(body.fields['body.numberMin5'].message).to.equal('min 5');
+        expect(body.fields['body.numberMin5'].value).to.equal(bodyModel.numberMin5);
+        expect(body.fields['body.stringMax10Lenght'].message).to.equal('maxLength 10');
+        expect(body.fields['body.stringMax10Lenght'].value).to.equal(bodyModel.stringMax10Lenght);
+        expect(body.fields['body.stringMin5Lenght'].message).to.equal('minLength 5');
+        expect(body.fields['body.stringMin5Lenght'].value).to.equal(bodyModel.stringMin5Lenght);
+        expect(body.fields['body.stringPatternAZaz'].message).to.equal('Not match in \'^[a-zA-Z]+$\'');
+        expect(body.fields['body.stringPatternAZaz'].value).to.equal(bodyModel.stringPatternAZaz);
+
+        expect(body.fields['body.arrayMax5Item'].message).to.equal('maxItems 5');
+        expect(body.fields['body.arrayMax5Item'].value).to.deep.equal(bodyModel.arrayMax5Item);
+        expect(body.fields['body.arrayMin2Item'].message).to.equal('minItems 2');
+        expect(body.fields['body.arrayMin2Item'].value).to.deep.equal(bodyModel.arrayMin2Item);
+        expect(body.fields['body.arrayUniqueItem'].message).to.equal('required unique array');
+        expect(body.fields['body.arrayUniqueItem'].value).to.deep.equal(bodyModel.arrayUniqueItem);
+        expect(body.fields['body.intersection'].message).to.deep.equal('Could not match the intersection against every type. Issues: [{"body.value2":{"message":"\'value2\' is required"}}]');
```

Comment:
@WoH these 4 lines are the only assertions I was able to find that verify that the validation code is working. I’d like to see some unit tests that execute the validation logic. I think they would be valuable because you could explain the rationale as to why the assertions should be true. The fact that you can add a “because” string as the second parameter of any Chai assertion is one of the few things that I like about Chai/Mocha (I generally prefer Jest).

So it would help me and future devs a great deal if there were tests like:
```
// pseudo-code
const result = validation(model, data, errorDictionary)

expect(result).to.equal(“something”, “because unions have this special trait that I’m going to explain now”)
```

Reply:
Another reason why unit tests are valuable for this PR is that you’ll want to test for the various cases for additionalProperties in the validation tests. Off the top of my head, the intersection validation might not work with what I wrote last week for additionalProperties validation. But that’s why we need some exhaustive unit tests.

Reply:
I am onto it. But in order to test some stuff, I discovered #432 (trying to verify 'string' | model)

Reply:
K, 432 has been merged.

Reply:
https://github.com/lukeautry/tsoa/pull/400/commits/ce1b89d5e03766c3e06b1a4f456a92d2184556cc

Reply:
Thank you for adding integrative tests. I should have been more clear that I was hoping for true unit tests that test one function at a time. So I've added a couple of unit tests right here in this comment. Please add these to `templateHelpers.spec.ts`.

```ts
        it('should allow additionalProperties (on a union) if noImplicitAdditionalProperties is set to silently-remove-extras', () => {
            // Arrange
            const refName = 'ExampleModel';
            const subSchemas: TsoaRoute.PropertySchema[] = [
                { "ref": "TypeAliasModel1" },
                { "ref": "TypeAliasModel2" }
            ]
            const models: TsoaRoute.Models = {
                [refName]:{
                    properties: {
                        or: {
                            "dataType": "union",
                            "subSchemas": subSchemas,
                            "required": true
                        },
                    }
                },
                "TypeAliasModel1": {
                    "properties": {
                      "value1": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
                "TypeAliasModel2": {
                    "properties": {
                      "value2": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
            }
            const v = new ValidationService(models);
            const minimalSwaggerConfig: SwaggerConfigRelatedToRoutes = {
                noImplicitAdditionalProperties: 'silently-remove-extras',
            };
            const errorDictionary: FieldErrors = {};
            const nameOfAdditionalProperty = 'I am the bad key name';
            let dataToValidate: {or: TypeAliasModel1} = {
                or: {
                    value1: "this is value 1"
                }
            };
            (dataToValidate.or as any)[nameOfAdditionalProperty]= 'something extra';

            // Act
            const name = 'dataToValidate';
            const result = v.validateUnion('or', dataToValidate, errorDictionary, minimalSwaggerConfig, subSchemas, name+".");

            // Assert
            expect(errorDictionary).to.deep.eq({});
            expect(result).to.eql({
                or: {
                    value1: "this is value 1",
                    [nameOfAdditionalProperty]: 'something extra',
                }
            });
            if (dataToValidate[nameOfAdditionalProperty]) {
                throw new Error(`dataToValidate.${nameOfAdditionalProperty} should have been removed because "silently-remove-extras" requires that excess properties be stripped.`);
            }
        });

        it('should throw if the data has additionalProperties (on a union) if noImplicitAdditionalProperties is set to throw-on-extras', () => {
            // Arrange
            const refName = 'ExampleModel';
            const subSchemas: TsoaRoute.PropertySchema[] = [
                { "ref": "TypeAliasModel1" },
                { "ref": "TypeAliasModel2" }
            ]
            const models: TsoaRoute.Models = {
                [refName]:{
                    properties: {
                        or: {
                            "dataType": "union",
                            "subSchemas": subSchemas,
                            "required": true
                        },
                    }
                },
                "TypeAliasModel1": {
                    "properties": {
                      "value1": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
                "TypeAliasModel2": {
                    "properties": {
                      "value2": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
            }
            const v = new ValidationService(models);
            const minimalSwaggerConfig: SwaggerConfigRelatedToRoutes = {
                noImplicitAdditionalProperties: 'throw-on-extras',
            };
            const errorDictionary: FieldErrors = {};
            const nameOfAdditionalProperty = 'I am the bad key name';
            let dataToValidate: TypeAliasModel1 = {
                value1: "valueOne"
            };
            (dataToValidate as any)[nameOfAdditionalProperty]= 'something extra';

            // Act
            const name = 'dataToValidate';
            v.validateUnion('or', dataToValidate, errorDictionary, minimalSwaggerConfig, subSchemas, name+".");

            // Assert
            const errorKeys = Object.keys(errorDictionary);
            expect(errorKeys).to.have.lengthOf(1);
            const firstAndOnlyErrorKey = errorKeys[0];
            expect(errorDictionary[firstAndOnlyErrorKey].message).to.include(`Could not match the union against any of the items.`);
            expect(errorDictionary[firstAndOnlyErrorKey].message).to.include(nameOfAdditionalProperty);
            expect(errorDictionary[firstAndOnlyErrorKey].message).to.include(`is an excess property and therefore is not allowed`);
            if (!dataToValidate[nameOfAdditionalProperty]) {
                throw new Error(`dataToValidate.${nameOfAdditionalProperty} should have been there because .validateModel should NOT have removed it since it took the more severe option of producing an error instead.`);
            }
        });

        it('should throw if the data has additionalProperties (on a intersection) if noImplicitAdditionalProperties is set to throw-on-extras', () => {
            // Arrange
            const refName = 'ExampleModel';
            const subSchemas: TsoaRoute.PropertySchema[] = [
                { "ref": "TypeAliasModel1" },
                { "ref": "TypeAliasModel2" }
            ]
            const models: TsoaRoute.Models = {
                [refName]:{
                    properties: {
                        and: {
                            "dataType": "intersection",
                            "subSchemas": subSchemas,
                            "required": true
                        },
                    }
                },
                "TypeAliasModel1": {
                    "properties": {
                      "value1": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
                "TypeAliasModel2": {
                    "properties": {
                      "value2": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
            }
            const v = new ValidationService(models);
            const minimalSwaggerConfig: SwaggerConfigRelatedToRoutes = {
                noImplicitAdditionalProperties: 'throw-on-extras',
            };
            const errorDictionary: FieldErrors = {};
            const nameOfAdditionalProperty = 'extraKeyName';
            const expectedErrorMsg = `The following properties are not allowed by any part of the intersection: ${nameOfAdditionalProperty}`;
            let dataToValidate: TypeAliasModel1 & TypeAliasModel2 = {
                value1: "this is value 1",
                value2: "this is value 2"
            };
            (dataToValidate as any)[nameOfAdditionalProperty]= 'something extra';

            // Act
            const name = 'dataToValidate';
            v.validateIntersection('and', dataToValidate, errorDictionary, minimalSwaggerConfig, subSchemas, name+".");

            // Assert
            const errorKeys = Object.keys(errorDictionary);
            expect(errorKeys).to.have.lengthOf(1);
            const firstAndOnlyErrorKey = errorKeys[0];
            expect(errorDictionary[firstAndOnlyErrorKey].message).to.eq(expectedErrorMsg);
            if (!dataToValidate[nameOfAdditionalProperty]) {
                throw new Error(`dataToValidate.${nameOfAdditionalProperty} should have been there because .validateModel should NOT have removed it since it took the more severe option of producing an error instead.`);
            }
        });

        it('should not throw if the data has additionalProperties (on a intersection) if noImplicitAdditionalProperties is set to silently-remove-extras', () => {
            // Arrange
            const refName = 'ExampleModel';
            const subSchemas: TsoaRoute.PropertySchema[] = [
                { "ref": "TypeAliasModel1" },
                { "ref": "TypeAliasModel2" }
            ]
            const models: TsoaRoute.Models = {
                [refName]:{
                    properties: {
                        and: {
                            "dataType": "intersection",
                            "subSchemas": subSchemas,
                            "required": true
                        },
                    }
                },
                "TypeAliasModel1": {
                    "properties": {
                      "value1": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
                "TypeAliasModel2": {
                    "properties": {
                      "value2": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
            }
            const v = new ValidationService(models);
            const minimalSwaggerConfig: SwaggerConfigRelatedToRoutes = {
                noImplicitAdditionalProperties: 'silently-remove-extras',
            };
            const errorDictionary: FieldErrors = {};
            const nameOfAdditionalProperty = 'extraKeyName';
            let dataToValidate: TypeAliasModel1 & TypeAliasModel2 = {
                value1: "this is value 1",
                value2: "this is value 2"
            };
            (dataToValidate as any)[nameOfAdditionalProperty]= 'something extra';

            // Act
            const name = 'dataToValidate';
            const result = v.validateIntersection('and', dataToValidate, errorDictionary, minimalSwaggerConfig, subSchemas, name+".");

            // Assert
            expect(errorDictionary).to.deep.eq({});
            if (dataToValidate[nameOfAdditionalProperty]) {
                throw new Error(`dataToValidate.${nameOfAdditionalProperty} should have been removed because "silently-remove-extras" requires that excess properties be stripped.`);
            }
            expect(result).to.eql({
                value1: "this is value 1",
                value2: "this is value 2",
            });
        });

        it('should not throw if the data has additionalProperties (on a intersection) if noImplicitAdditionalProperties is set to false', () => {
            // Arrange
            const refName = 'ExampleModel';
            const subSchemas: TsoaRoute.PropertySchema[] = [
                { "ref": "TypeAliasModel1" },
                { "ref": "TypeAliasModel2" }
            ]
            const models: TsoaRoute.Models = {
                [refName]:{
                    properties: {
                        and: {
                            "dataType": "intersection",
                            "subSchemas": subSchemas,
                            "required": true
                        },
                    }
                },
                "TypeAliasModel1": {
                    "properties": {
                      "value1": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
                "TypeAliasModel2": {
                    "properties": {
                      "value2": { "dataType": "string", "required": true },
                    },
                    "additionalProperties": false,
                },
            }
            const v = new ValidationService(models);
            const minimalSwaggerConfig: SwaggerConfigRelatedToRoutes = {
                noImplicitAdditionalProperties: false,
            };
            const errorDictionary: FieldErrors = {};
            const nameOfAdditionalProperty = 'extraKeyName';
            let dataToValidate: TypeAliasModel1 & TypeAliasModel2 = {
                value1: "this is value 1",
                value2: "this is value 2"
            };
            (dataToValidate as any)[nameOfAdditionalProperty]= 'something extra';

            // Act
            const name = 'dataToValidate';
            const result = v.validateIntersection('and', dataToValidate, errorDictionary, minimalSwaggerConfig, subSchemas, name+".");

            // Assert
            expect(errorDictionary).to.deep.eq({});
            expect(result).to.eql({
                value1: "this is value 1",
                value2: "this is value 2",
                [nameOfAdditionalProperty]: 'something extra',
            });
        });
```

1 of those tests won't pass. I think it's because it's incredibly challenging to match an object against a union of objects. So I might be okay with changing `should allow additionalProperties (on a union) if noImplicitAdditionalProperties is set to silently-remove-extras` to instead assert `should not allow additionalProperties (on a union) regardless of the noImplicitAdditionalProperties config due to the complexity of unions`.

Two of the other tests won't pass until you fix this other part that I gave the solution for in https://github.com/lukeautry/tsoa/pull/400/files#r316392569

After these unit tests are added and the other comments I've made are resolved, we can merge this PR. :)